### PR TITLE
feat: Advanced logic editor

### PR DESCRIPTION
**Pull request by gupta-piyush19**
## What does this PR do?
This pull request includes significant updates to the logic editor documentation, image format changes, navigation updates, survey template refactoring, and component imports. The most important changes include adding a new Logic Editor documentation page, converting image formats from PNG to WEBP, updating the navigation to include the new Logic Editor, refactoring survey templates to use functions, and updating the `AdvancedSettings` component to use `ConditionalLogic`.

### Documentation Updates:
* [`apps/docs/app/global/logic-editor/page.mdx`](diffhunk://#diff-da54eb0ee919d657baeffa22456572ab7ae795232cd8b8afc2085d4ec995a504R1-R171): Added a comprehensive documentation page for the Logic Editor, including images and detailed instructions on creating and managing logic blocks.

### Image Format Changes:
* [`apps/docs/app/self-hosting/configuration/page.mdx`](diffhunk://#diff-cbb84e6649d84f74844ffe230e603a283cf0ef49ffe9bfd192c9d079a24603e2L3-R12): Converted image formats from PNG to WEBP for better performance.

### Navigation Updates:
* [`apps/docs/lib/navigation.ts`](diffhunk://#diff-4d393ffcdbdfc3286e285cbc05e8dfb5fb0165fc49d62650f39f718884f5c5e3R112): Added a new navigation link for the Logic Editor documentation.

### Survey Template Refactoring:
* `apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.ts`: Refactored survey templates (`NPSSurvey`, `StarRatingSurvey`, `CSATSurvey`, `CESSurvey`, `SmileysRatingSurvey`, `eNPSSurvey`) to use functions, improving code maintainability and reusability. ([apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.tsL15-R16](diffhunk://#diff-d9f51f9f641a7965615bf6a0bbe7dc07c1d3c054c83004dcfa25ebc58a0c8bb7L15-R16), [apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.tsR45-R86](diffhunk://#diff-d9f51f9f641a7965615bf6a0bbe7dc07c1d3c054c83004dcfa25ebc58a0c8bb7R45-R86), [apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.tsL62-R132](diffhunk://#diff-d9f51f9f641a7965615bf6a0bbe7dc07c1d3c054c83004dcfa25ebc58a0c8bb7L62-R132), [apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.tsR143-R184](diffhunk://#diff-d9f51f9f641a7965615bf6a0bbe7dc07c1d3c054c83004dcfa25ebc58a0c8bb7R143-R184), [apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.tsL102-R228](diffhunk://#diff-d9f51f9f641a7965615bf6a0bbe7dc07c1d3c054c83004dcfa25ebc58a0c8bb7L102-R228), [apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.tsR237-R240](diffhunk://#diff-d9f51f9f641a7965615bf6a0bbe7dc07c1d3c054c83004dcfa25ebc58a0c8bb7R237-R240), [apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.tsR265-R306](diffhunk://#diff-d9f51f9f641a7965615bf6a0bbe7dc07c1d3c054c83004dcfa25ebc58a0c8bb7R265-R306), [apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.tsL164-R352](diffhunk://#diff-d9f51f9f641a7965615bf6a0bbe7dc07c1d3c054c83004dcfa25ebc58a0c8bb7L164-R352), [apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.tsR363-R366](diffhunk://#diff-d9f51f9f641a7965615bf6a0bbe7dc07c1d3c054c83004dcfa25ebc58a0c8bb7R363-R366), [apps/web/app/(app)/(onboarding)/environments/[environmentId]/xm-templates/lib/xm-templates.tsR397-R405](diffhunk://#diff-d9f51f9f641a7965615bf6a0bbe7dc07c1d3c054c83004dcfa25ebc58a0c8bb7R397-R405))

### Component Import Updates:
* `apps/web/app/(app)/(survey-editor)/environments/[environmentId]/surveys/[surveyId]/edit/components/AdvancedSettings.tsx`: Updated the `AdvancedSettings` component to use `ConditionalLogic` instead of `LogicEditor`. ([apps/web/app/(app)/(survey-editor)/environments/[environmentId]/surveys/[surveyId]/edit/components/AdvancedSettings.tsxR1-L3](diffhunk://#diff-8df395275e9acd5e3a5c20626a9819e3cf88414ed65108022c8c30526a2cbf75R1-L3), [apps/web/app/(app)/(survey-editor)/environments/[environmentId]/surveys/[surveyId]/edit/components/AdvancedSettings.tsxL22-L31](diffhunk://#diff-8df395275e9acd5e3a5c20626a9819e3cf88414ed65108022c8c30526a2cbf75L22-L31))

Fixes 
https://github.com/formbricks/formbricks/issues/3059
https://github.com/formbricks/internal/issues/273
https://github.com/formbricks/internal/issues/312

and also [#2691](https://github.com/formbricks/formbricks/issues/2691)

### CODE REVIEW

**Comment by jobenjada** [comment]
hey Piyush, I'll pre-review this to save us a feedback loop 🤓 

It's looking great already! Here's what I noticed?

1. Let's move this inside of the last block of Conditional Logic. Some users created logic for all options because they didn't see it. If it sits inside the gray box, it should be clearer. Also, make sure the font size and color is the same like the "When" and "Then" texts in the gray container:
image

2. Text size: The rest of the surey editor is `sm` so please change all of the texts to sm within the logic editor
image

2.b. Icon size and color: These icons can be a bit smaller and darker for better usability
image

3. Keep spacing between input fields consistent: In the conditions its too wide, in the actions too narrow. Maybe `gap-x-2` should work
image

also gap here is too wide and inconsistent with gap between Condition and Action
image

I think you can get rid of `mt-2` here
image

4. hover:cursor-pointer
image

5. Remove the outer gray container: It's clear enough and gives us some space
image

6. Replace Emoji with lucide `file-digit`
image

here as well:
image

7. Tweak Combobox UI: border-r not rounded if item selected
image

and there is a double border separating the input from the dropdown
image

8. Spacing: Remove top padding
image

8.b. This icon needs more space on the right:
image


9. Logic block: Since they are executed in sync, we need to allow resorting them. Replace delete icon with context menu for logic block with: Duplicate, Move Up, Move Down, and Delete. Add the icons we use for this functionality on Question Cards
image

10 Update Delete icon to match rest of survey editor:
image

vs

image

11.  I think this needs to get filtered out? Not sure if that makes sense
image

That's it in terms of UI / UX. I'll do a proper test when you've marked the PR ready for review 🤓 

**Thanks, great work!!**

**Comment by jobenjada** [comment]
Please make sure to add the variables to the response data (exports, integrations, etc)

**Review by Dhruwang — CHANGES_REQUESTED** [review]
Hey @gupta-piyush19 , tested the logic functionality, overall works well 🚀🙌🏻

Here are some issues that i noticed.

### Modifying logic operator in open text question does not work 

Steps to reproduce  

1. Add a logic to an open text question 
2. Change operator (for example change `equals` to `does not equal`)

Noticed the same issue in date question (please check for other questions as well)  

https://github.com/user-attachments/assets/f25414c7-0a3f-48d9-b1b4-303952385b0b

Screenshot 2024-09-10 at 2 14 22 PM

 
 
### Cannot save survey with logic where optional question is logically set to require
 
Steps to reproduce 
1. Add a logic to a question where action is `Require Answer` 
2. Select an optional question from the dropdown 
3. Try to save the survey    


https://github.com/user-attachments/assets/5dd4b4f6-58f3-4aa7-bfe4-f752941dcd6c



### Questions in operand dropdown 

In SS, it can be seen that for open text questions, operand input is a dropdown. So is it also possible to compare value of current question with answer of those question included in the dropdown?  or that dropdown should not contain questions  🤔
Screenshot 2024-09-10 at 10 10 25 AM


### Fix alignment of icon on variables card 
Screenshot 2024-09-10 at 10 30 17 AM



### Jumps based on variable values does not work  

As you can see in the video below, I have a variable named `dhruwang`, which has an initial value of 0. Then I set a condition in first question that if value of variable `dhruwang` is 0, then jump to Thank you card. But is does not seem to work  


https://github.com/user-attachments/assets/fa486cda-d0db-4e33-a0d9-69840668cc93



### Not able to save survey, when logic of some question contain an action which includes calculate operation on text variables  

https://github.com/user-attachments/assets/71cae7f6-83dd-421f-9910-13a377f41298


### We cannot make a question optional logically, So we can remove `is Skipped` from the dropdown for required questions   

Screenshot 2024-09-10 at 11 23 17 AM


### For multiple choice multi select question, if operator is `equals`, then user should be able to select multiple operands as we did in older logic (because response to multiple choice multiple select is an array)  

https://github.com/user-attachments/assets/01005529-625e-4b35-9798-1b48d8915eca




### is Submitted and is Skipped not working for multiple choice questions  

### Equal (=) and not equals (!=) not working for NPS and rating questions  


https://github.com/user-attachments/assets/c35e9fbe-bb33-4074-b5e7-da8e080f869c



### Issue with multiple conditions  

As you can see in the video below, I set two conditions joined with `and` which says, that if the open text question is skipped and rating is above 5 then show the ending card. But it shows the ending card even if open text question is filled  

  

https://github.com/user-attachments/assets/196309eb-38ac-47d8-8541-b351afa80117



### image preview in picture select logic settings   

Is is possible to show image preview in picture select logic dropdown instead of image name, similar to what we did in older logic editor   

Screenshot 2024-09-10 at 11 56 11 AM

Screenshot 2024-09-10 at 11 57 25 AM

### Logic in ranking question not working   

Screenshot 2024-09-10 at 1 18 24 PM



### We can now remove `is partially submitted` from matrix question as we do not allow partial submissions now   
Screenshot 2024-09-10 at 1 20 58 PM



### `is Skipped` in file upload not working  

https://github.com/user-attachments/assets/e02cdc81-7508-4190-8dba-572c07c1a0eb


### Can you also recheck logic evaluation using hidden field values 


Can you check these things ? 😊

**Review by pandeymangg — CHANGES_REQUESTED** [review]
thanks a lot for this biiiiiig PR @gupta-piyush19 great work on the code overall! 
I just had a couple of changes suggested mainly about consistency, pls take a look.
Found these issues in the UI for now, I'll test it more:

The text in the toast overflows if you have a logic error in a question that also uses recall, we also somehow leak this recall message
image

Also found this issue that changing the question headline rapidly, the survey editor goes into the infinite render error when the logic editor is open

https://github.com/user-attachments/assets/d568d0f0-ca4e-4542-a793-8c5405b6ce8a

File: packages/types/surveys/types.ts [hunk 1]
```diff
@@ -1066,6 +896,826 @@ export const ZSurvey = z
     });
   });
 
+const isInvalidOperatorsForQuestionType = (
```
[the remark this answers is not in the document]
Comment (gupta-piyush19) [reply packages/types/surveys/types.ts:898]: getting circular dependency issue in this change

**Comment by pandeymangg** [comment]
found a couple of more issues while testing:

1. The full recall string is visible here if a question contains recall:

image

2. The variable `score` here is a number and is equal to `0` but in the recall, it renders the fallback value, I think its treating the `0` as a `false` value, which shouldn't be the case in terms of number variables:

image

3. In this single select question, both the logic blocks contain the same logic (same condition and same action), but it only takes the last logic block (the second one in this case)

image

**Comment by jobenjada** [comment]
Overall, really cool! I've made a few tweaks in an effort to unify our dropdown menus (Segments, Logic, and rest of app)

as discussed:
image

spacing above Conditional Logic too high
image

lets add a bit of padding above the scroll bar
image

increase width of toast globally (this can be wider in general if the toast message requires it. maybe `w-fit max-w-md` )
image

and lets talk about performance on Slack 🤙

**Review by Dhruwang — CHANGES_REQUESTED** [review]
Logic for all question works really well, just noticed two issues that needs to be addressed 😊

When open text is configured to take number input, skipped condition is not working (works for strings)

https://github.com/user-attachments/assets/1ef70da3-5c54-4420-9de6-cf7f7bacbac4



Similar kind of issue with multiple select single


https://github.com/user-attachments/assets/8f334ff3-9664-47bc-963f-8c0a557c9383

**Review by pandeymangg — APPROVED** [review]

**Pull request merged**